Expand the Access Controls reference

ptgott opened the pull request:
Closes #11263

Teleport roles make use of separate but similar predicate languages.
Clarify these languages by giving each one a section within the Access
Controls reference.

This also clarifies other aspects of predicates in Teleport roles:

- Explain the dot vs. bracket syntax in `external` variables
- Add a note re: OIDC claim data types

To make room for the new organization, this also changes other aspects
of the page's structure:
- Some paragraphs in the introductory section were more appropriate in
  the section re: managing access to Teleport RBAC resources.
- Make the intro and title more readable.
- Move an out-of-place H3 regarding SAML to the `external` trait
  section.

Commit:
add template vars/functions reference

Commit:
Update docs/pages/access-controls/reference.mdx

Commit:
replace details blocks with headers

With unopened details blocks the headers cannot be directly linked to, negating their usefullness as H4s

Commit:
partial updates

ptgott committed:
Clarify predicate language contexts

Teleport roles make use of different but similar predicate languages.
Clarify these languages by giving each one a section within the Access
Controls reference.

This also clarifies other aspects of predicates in Teleport roles:

- Explain the dot vs. bracket syntax in `external` variables
- Add a note re: OIDC claim data types

To make room for the new organization, this also changes other aspects
of the page's structure:
- Some paragraphs in the introductory section were more appropriate in
  the section re: managing access to Teleport RBAC resources.
- Make the intro and title more readable.
- Move an out-of-place H3 regarding SAML to the `external` trait
  section.

ptgott commented:
One thing we should determine is whether the edits subsume the Role Templates guide (docs/pages/access-controls/guides/role-templates.mdx) enough to merit deleting that guide, which is less detailed.

lsgunn-teleport commented on docs/pages/access-controls/reference.mdx:
```diff
@@ -1,101 +1,46 @@
-To manage cluster roles, a Teleport administrator can use the Web UI or the
-command line using [tctl resource commands](../reference/resources.mdx).
-To see the list of roles in a Teleport cluster, an administrator can execute:
+To manage cluster roles, as well as other dynamic resources, a Teleport
+administrator can use the Web UI or a [Teleport client
+tool](../management/dynamic-resources.mdx) like `tctl`, the Teleport Terraform
+provider, or the Teleport Kubernetes operator, as well as [custom API
+clients](../api/rbac.mdx).
```
```suggestion
You can any of the following to manage cluster roles and other dynamic resources:

- Teleport Web UI
- [Teleport clienttool](../management/dynamic-resources.mdx) like `tctl`
- Teleport Terraform provider
- Teleport Kubernetes operator
- [custom API clients](../api/rbac.mdx)
```
I'm not sure this feels particularly natural here, but the two "as well as" clauses seemed a little tricky to pull off.

lsgunn-teleport commented on docs/pages/access-controls/reference.mdx:
```diff
@@ -418,14 +320,253 @@ spec:
+<Admonition type="warning">
+Label expressions are available starting in Teleport version `13.1.1`.
+All components of your Teleport cluster must be upgraded to version `13.1.1`
+or newer before you will be able to use label expressions.
```
```suggestion
or newer before you can use label expressions.
```

lsgunn-teleport commented on docs/pages/access-controls/reference.mdx:
```diff
@@ -418,14 +320,253 @@ spec:
+
+### Label expressions
+
+<Admonition type="warning">
```
```suggestion
```
Does this really warrant a Warning?

zmb3 replied:
Because it's so new, I would say yes. (at least for the v13 and v12 versions of the docs)

If you don't see this, and you create a label expression on too old of a version, you will think you are enforcing security properties when you are not. This could lead to unintended infrastructure access, which would be a critical incident.

lsgunn-teleport commented on docs/pages/access-controls/reference.mdx:
````diff
@@ -418,14 +320,253 @@ spec:
+      contains(user.spec.traits["teams"], labels["team"])
+```
+
+Typically you will only want to use one of `<kind>_labels` or
````
```suggestion
Typically, you would only use either `<kind>_labels` or
```

lsgunn-teleport commented on docs/pages/access-controls/reference.mdx:
```diff
@@ -418,14 +320,253 @@ spec:
+Typically you will only want to use one of `<kind>_labels` or
+`<kind>_labels_expression` in a single role, but you are allowed to specify
+both.
+If you have both in a deny rule, the matching is greedy, if either one matches
+access will be denied.
```
```suggestion
If you have both in a deny rule, the matching is greedy. If either expression matches,
access is denied.
```

lsgunn-teleport commented on docs/pages/access-controls/reference.mdx:
```diff
@@ -418,14 +320,253 @@ spec:
+both.
+If you have both in a deny rule, the matching is greedy, if either one matches
+access will be denied.
+In an allow rule, the matching is not greedy, if both are set they both have to
+match for access to be allowed.
```
```suggestion
In an allow rule, the matching is not greedy. If both expressions are set, both expressions must be
matched for access to be allowed.
```

lsgunn-teleport commented on docs/pages/access-controls/reference.mdx:
```diff
@@ -418,14 +320,253 @@ spec:
+|`{{strings.upper(list)}}`|Returns the new list of strings that results from uppercasing each item in `list`, a list of strings.|
+|`{{strings.lower(list)}}`|Returns the new list of strings that results from lowercasing each item in `list`, a list of strings.|
+
+### Predicates for access to infrastructure resources
```
```suggestion
### Predicates for access to infrastructure resources
```
Might just be my ignorance, but I don't see the connection between this heading and the paragraph and bullets that follow.

nklaassen replied:
I don't think of the below as predicates. I'll borrow a wikipedia definition for predicate:
> A predicate is a statement or mathematical assertion that contains variables, sometimes referred to as predicate variables, and may be true or false depending on those variables’ value or values.

Label expressions and search filters *are* predicates, because they evaluate to true or false. The below like `logins: "{{internal.logins}}"` is not a predicate that is true or false. It is a template that can be populated with the user traits to produce a value, like `logins: ["nic", "root"]`

lsgunn-teleport commented on docs/pages/access-controls/reference.mdx:
```diff
@@ -418,14 +320,253 @@ spec:
+
+You can use the following functions in role fields that govern access to
+principals on external systems. Both functions can take a template variable
+named above as an argument:
```
```suggestion
as an argument:
```

ptgott replied:
Removed this sentence, since the list of variables is below the list of functions.

lsgunn-teleport commented on docs/pages/access-controls/reference.mdx:
```diff
@@ -418,14 +320,253 @@ spec:
+| Variable | Description |
+| - | - |
+| `{{internal.logins}}` | Substituted with a value stored in Teleport's local user database and logins from a root cluster. <br/><br/>For local users, Teleport will substitute this with the "allowed logins" parameter used in the `tctl users add [user] <allowed logins>` command. <br/><br/>If the role is within a leaf cluster in a [trusted cluster](../management/admin/trustedclusters.mdx), Teleport ...
+|`{{internal.azure_identities}}`| Returns the Azure identities defined in Teleport available to the user. Azure identities can set for a specific user or defined in a role.|
```
```suggestion
|`{{internal.azure_identities}}`| Returns the Azure identities defined in Teleport available to the user. Azure identities can be set for a specific user or defined in a role.|
```

lsgunn-teleport commented on docs/pages/access-controls/reference.mdx:
```diff
@@ -418,14 +320,253 @@ spec:
+|`{{internal.kubernetes_users}}`|List of allowed Kubernetes users for the user.|
+|`{{internal.logins}}`|List of allowed SSH logins for the user.|
+|`{{internal.windows_logins}}`|List of allowed Windows logins for the user.|
+| `{{external.xyz}}` | Substituted with a value from an external [SSO provider](https://en.wikipedia.org/wiki/Single_sign-on). If using SAML, this will be expanded with "xyz" assertion value. For OIDC, this will be expanded a value of "xyz" claim. See the next section for more information on referring to the `external` variable in Teleport roles. |
```
```suggestion
| `{{external.xyz}}` | Substituted with a value from an external [SSO provider](https://en.wikipedia.org/wiki/Single_sign-on). If using SAML, this variable is expanded with the "xyz" assertion value. For OIDC, the variable is expanded to the value of the "xyz" claim. See the next section for more information on referring to the `external` variable in Teleport roles. |
```

lsgunn-teleport commented on docs/pages/access-controls/reference.mdx:
```diff
@@ -418,14 +320,253 @@ spec:
+- **Dot syntax:** `{{external.trait}}`
+- **Bracket syntax:** `{{external["trait"]}}`
+
+Teleport will expand the template variables above with the value of the SAML
+attribute or OIDC claim called `trait`.
```
```suggestion
Teleport expands the template variables with the value of the SAML
attribute or OIDC claim called `trait`.
```

lsgunn-teleport reviewed (approved):
Some minor suggestions and questions.

zmb3 commented on docs/pages/access-controls/reference.mdx:
````diff
@@ -418,14 +320,253 @@ spec:
 ```
 
-## Second Factor - U2F
+## Role predicates
````
Should we call this "role templates" instead?

zmb3 commented on docs/pages/access-controls/reference.mdx:
```diff
@@ -418,14 +320,253 @@ spec:
+|`{{regexp.match(regexp)}}`|Returns `true` if the regular expression matches a user's role.|
+|`{{regexp.not_match(regexp)}}`|Returns `true` if the regular expression does not match a user's role.|
+
+Regular expressions support both wildcards (the `*` character) and [Go-style
```
```suggestion
Regular expressions support both glob-style wildcards (the `*` character) and [Go-style
```

zmb3 commented on docs/pages/access-controls/reference.mdx:
```diff
@@ -418,14 +320,253 @@ spec:
+#### Fields
+
+The `<kind>_labels_expression` fields have the same purpose as the matching
+`<kind>_labels` fields, but support predicate expressions instead of label
```
```suggestion
`<kind>_labels` fields, but support label expressions instead of label
```

I still think we need to call these by their proper name. Label expressions are different than the generic predicate language used to search and filter resources.

zmb3 commented on docs/pages/access-controls/reference.mdx:
```diff
@@ -418,14 +320,253 @@ spec:
+
+When a user attempts to authenticate to an infrastructure resource
+proxied by Teleport, such as a database or Kubernetes cluster, Teleport determines from the user's roles:
+- Which infrastructure resources the user is authorized to authenticate to. 
```
```suggestion
- Whether the user is authorized to connect
```

It seems odd to say "when attempting to connect to a specific resource Teleport determines all possible resources the user is allowed to connect to"

zmb3 commented on docs/pages/access-controls/reference.mdx:
```diff
@@ -418,14 +320,253 @@ spec:
+- `node_labels`
+- `windows_desktop_labels`
+
+Principals a user can assume on third-party systems:
```
+1 I'm not sure either

ptgott replied:
Is "infrastructure resources" clearer?

zmb3 commented on docs/pages/access-controls/reference.mdx:
```diff
@@ -418,14 +320,253 @@ spec:
+role resource.
+
+Labels for resources enrolled with Teleport:
+- `app_labels`
```
I think we should introduce `app_labels` before `app_labels_expression` as it is simpler and much more common in practice.

zmb3 commented on docs/pages/access-controls/reference.mdx:
```diff
@@ -418,14 +320,253 @@ spec:
+|`{{email.local(variable)}}`|Extracts the local part of an email address.|
+|`{{regexp.replace(variable, expression, replacement)}}`|Finds all matches of the expression within the variable and replaces them with the replacement.|
+
+#### Variables
```
These are typically referred to as "traits"

nklaassen commented on docs/pages/access-controls/reference.mdx:
```diff
@@ -418,14 +320,253 @@ spec:
+
+#### Variables
+
+You can use the following template variables in label expression fields:
+
+|Variable|Type|Description|
+|---|---|---|
+|`{{user.spec.traits}}`|`map[string][]string`|All of the user's traits. Use `{{user.spec.traits<key>}}` to get a list of trait values for the specified key.|
+|`{{labels}}`|`map[string]string`|All of the user's labels. Use `{{labels.<key>}}` to get the value for the specified key.|
+
+#### Functions
+
+You can use the following functions in label expression fields:
+
+|Function|Description|
+|---|---|
+|`{{contains(list,item)}}`|Returns `true` if the list of strings in `list` contains the specified `item` and `false` otherwise.|
+|`{{regexp.match(list,regexp)}}`|Returns `true` if the regular expression in `regexp` matches any of the strings in list of strings `list`. Returns `false` otherwise.|
+|`{{regexp.replace(list,regexp,replacement)}}`|Checks the list of strings `list`, returning a new list where each item that matches `regexp` is replaced with `replacement`. You can use `$n` in the replacement string to indicate a submatch group, e.g., `$1` to indicate the first submatch.|
+|`{{strings.upper(list)}}`|Returns the new list of strings that results from uppercasing each item in `list`, a list of strings.|
+|`{{strings.lower(list)}}`|Returns the new list of strings that results from lowercasing each item in `list`, a list of strings.|
```
Label expressions are different, they are not templates, they do not use the double-brace (`{{}}`) syntax.

With role templates, all templated fields with `{{}}` are populated at once based on the user traits to output the final role for the user. Label expressions are not templated, they are always expressions, and are evaluated at runtime when making access decisions. They take as input both the user's traits and the labels of a specific resource they are currently trying to access.

ptgott replied:
Thanks! I'll restore the "Label expressions" section to the way it was before this PR.

nklaassen commented on docs/pages/access-controls/reference.mdx:
```diff
@@ -418,14 +320,253 @@ spec:
+
+|Variable|Type|Description|
+|---|---|---|
+|`{{user.spec.traits}}`|`map[string][]string`|All of the user's traits. Use `{{user.spec.traits<key>}}` to get a list of trait values for the specified key.|
```
```suggestion
|`user.spec.traits`|`map[string][]string`|All of the user's traits. Use `user.spec.traits[<key>]` to get a list of trait values for the specified key.|
```

ptgott replied:
Restoring the "Label expressions" section to the way it was before this PR.

nklaassen commented on docs/pages/access-controls/reference.mdx:
```diff
@@ -418,14 +320,253 @@ spec:
+|Variable|Type|Description|
+|---|---|---|
+|`{{user.spec.traits}}`|`map[string][]string`|All of the user's traits. Use `{{user.spec.traits<key>}}` to get a list of trait values for the specified key.|
+|`{{labels}}`|`map[string]string`|All of the user's labels. Use `{{labels.<key>}}` to get the value for the specified key.|
```
```suggestion
|`labels`|`map[string]string`| Combined static and dynamic labels of the resource (server, application, etc.) being accessed. Use `labels[key]` to get the value for the specified key.|
```

ptgott replied:
Restoring the "Label expressions" section to the way it was before this PR.

nklaassen commented on docs/pages/access-controls/reference.mdx:
```diff
@@ -418,14 +320,253 @@ spec:
+|---|---|
+|`{{contains(list,item)}}`|Returns `true` if the list of strings in `list` contains the specified `item` and `false` otherwise.|
+|`{{regexp.match(list,regexp)}}`|Returns `true` if the regular expression in `regexp` matches any of the strings in list of strings `list`. Returns `false` otherwise.|
+|`{{regexp.replace(list,regexp,replacement)}}`|Checks the list of strings `list`, returning a new list where each item that matches `regexp` is replaced with `replacement`. You can use `$n` in the replacement string to indicate a submatch group, e.g., `$1` to indicate the first submatch.|
```
I think the text I currently have on the predicate page is more accurate
> Replaces all matches of `re` with `replacement` for all items in `list`

Because each item in the list might have one or more match for `re`, and each match will be replaced within that item. Currently it sounds like the whole item will be replaced.

nklaassen commented on docs/pages/access-controls/reference.mdx:
```diff
@@ -418,14 +320,253 @@ spec:
+
+#### Variables
+
+You can use the following template variables in label expression fields:
```
```suggestion
You can use the following variables in label expression fields:
```

nklaassen commented on docs/pages/access-controls/reference.mdx:
```diff
@@ -418,14 +320,253 @@ spec:
+
+|Function|Description|
+|---|---|
+|`{{contains(list,item)}}`|Returns `true` if the list of strings in `list` contains the specified `item` and `false` otherwise.|
+|`{{regexp.match(list,regexp)}}`|Returns `true` if the regular expression in `regexp` matches any of the strings in list of strings `list`. Returns `false` otherwise.|
+|`{{regexp.replace(list,regexp,replacement)}}`|Checks the list of strings `list`, returning a new list where each item that matches `regexp` is replaced with `replacement`. You can use `$n` in the replacement string to indicate a submatch group, e.g., `$1` to indicate the first submatch.|
+|`{{strings.upper(list)}}`|Returns the new list of strings that results from uppercasing each item in `list`, a list of strings.|
+|`{{strings.lower(list)}}`|Returns the new list of strings that results from lowercasing each item in `list`, a list of strings.|
```
```suggestion
|`contains(list,item)`|Returns `true` if the list of strings in `list` contains the specified `item` and `false` otherwise.|
|`regexp.match(list,regexp)`|Returns `true` if the regular expression in `regexp` matches any of the strings in list of strings `list`. Returns `false` otherwise.|
|`regexp.replace(list,regexp,replacement)`|Checks the list of strings `list`, returning a new list where each item that matches `regexp` is replaced with `replacement`. You can use `$n` in the replacement string to indicate a submatch group, e.g., `$1` to indicate the first submatch.|
|`strings.upper(list)`|Returns the new list of strings that results from uppercasing each item in `list`, a list of strings.|
|`strings.lower(list)`|Returns the new list of strings that results from lowercasing each item in `list`, a list of strings.|
```

ptgott committed:
Respond to lsgunn-teleport feedback

- Readability improvements.

ptgott commented:
TODO:
- [ ] Respond to Lisa feedback
- [ ] Respond to Zac feedback
- [ ] Respond to Nic feedback

ptgott committed:
Respond to zmb3 feedback

- Replace "predicate" uses with more suitable terms.
- Clarify wording.
- Move the section that explains `app_labels` below the one that
  explains `app_labels_expression`.

ptgott committed:
Respond to nklaassen feedback

Restore the original label expressions H3

ptgott commented:
@nklaassen Just checking to see if you are planning to look at this again before I merge. Thanks!

ptgott committed:
Linter fixes

ravicious commented on docs/pages/access-controls/reference.mdx:
```diff
@@ -450,14 +416,176 @@ spec:
+
+- `request.search_as_roles`
+- `request.roles`
+- `review_requests.preview_as_roles`
```
`regexp.match` does not work with `preview_as_roles` and I think it probably doesn't work with `search_as_roles` either.

The tests show it working only with thresholds filter and `review_requests.where`.

https://github.com/gravitational/teleport/blob/b05a3a79864cc1e709025d183b8a3663f90f13ea/lib/services/access_request_test.go#L242-L248

https://github.com/gravitational/teleport/blob/b05a3a79864cc1e709025d183b8a3663f90f13ea/lib/services/access_request_test.go#L274-L277
